azure-functions-extension-base/azure/functions/extension/base/web.py

```diff
@@ -0,0 +1,127 @@
+        return new_class
+
+    @classmethod
+    def get_module(cls):
+        return cls._module
+
+    @classmethod
+    def module_imported(cls):
+        return cls._module is not None
+
+
+class RequestTrackerMeta(type):
+    _request_type = None
+
+    def __new__(cls, name, bases, dct, **kwargs):
+        new_class = super().__new__(cls, name, bases, dct)
+
+        request_type = dct.get('request_type')
+
+        if request_type is None:
+            raise Exception(f'Request type not provided for class {name}')
+
+        if cls._request_type is not None and cls._request_type != request_type:
+            raise Exception(f'Only one request type shall be recorded for class {name} '
+                            f'but found {cls._request_type} and {request_type}')
+        cls._request_type = request_type
+
+        return new_class
+
+    @classmethod
+    def get_request_type(cls):
+        return cls._request_type
+
+    @classmethod
+    def check_type(cls, pytype: type) -> bool:
+        return cls._request_type is not None and issubclass(pytype, cls._request_type)
+
+
+class ResponseTrackerMeta(type):
+    _response_types = {}
+
+    def __new__(cls, name, bases, dct, **kwargs):
+        new_class = super().__new__(cls, name, bases, dct)
+
+        label = dct.get('label')
+        response_type = dct.get('response_type')
+
+        if label is None:
+            raise Exception(f'Response label not provided for class {name}')
+        if response_type is None:
+            raise Exception(f'Response type not provided for class {name}')
+        if cls._response_types.get(label) is not None and cls._response_types.get(label) != response_type:
+            raise Exception(f'Only one response type shall be recorded for class {name} '
+                            f'but found {cls._response_types.get(label)} and {response_type}')
+
+        cls._response_types[label] = response_type
+
+        return new_class
+
+    @classmethod
+    def get_standard_response_type(cls):
+        return cls.get_response_type(ResponseLabels.STANDARD)
+
+    @classmethod
+    def get_response_type(cls, label):
+        return cls._response_types.get(label)
+
+    @classmethod
+    def check_type(cls, pytype: type) -> bool:
+        return cls._response_types is not None and any(issubclass(pytype, response_type)
+                                                        for response_type in cls._response_types.values())
+
+class WebApp(metaclass=ModuleTrackerMeta):
+    @abstractmethod
+    def route(self, func: Callable):
+        pass
+
+    @abstractmethod
+    def get_app(self):
+        pass
+
+
+class WebServer(metaclass=ModuleTrackerMeta):
+    def __init__(self, hostname, port, web_app: WebApp):
+        self.hostname = hostname
+        self.port = port
+        self.web_app = web_app.get_app()
+
+    @abstractmethod
+    async def serve(self):
+        pass
+
+
+def http_v2_enabled() -> bool:
```

Comment:
Shouldn't this method be a part of WebApp?

Reply:
will put into a state class